Sendgrid adapter

stefan-prokop-cz opened the pull request:
* This PR removes "old" files and dirs from `master` branch.
* Adds new Sendgrid adapter with agreed interface
* Adds tests
* Updates docs / readme
* Upgrades deps

smoliji commented:
from me just to justify the two methods of sending a message - via global `send` and `adapter.sender`

1) `send` is the real deal - if you have a library its sole purpose is to send messages via various channels, you expect it to have such a method and it should be the first thing you stumble upon - and it requires adapter and a message. This leads you to following step
2) Adapter. To use an adapter (FCM, whatever), first logical step is to create an instance necesary to make it working - hand in API key, secret, let it estabilish connectino to remote server..
3) you are ready to send. BUT just to be more effective in some cases - you dont want to pass adapters around, most of the time you just want a channel only and it sends it directly to a channel, hence you can bound send to the adapter - and we have `adapter.sender` and thats it.

grissius commented:
> 3\. BUT just to be more effective in some cases - you dont want to pass adapters around

All I tried to say was that this is the 98%. You initialize channel and send. `send` the "real deal" still requires the channel instance which has the method so it does not really bring any advantages.

I understand the bottom-up construction, but if there is no practical advantage to using one over the other and we all agree that the adapter method is more convenient, why do we need to expose the general `send` method?

If there is a sensible reason to use `send` directly, explain. If not, I would suggest not including it in the public API. Keeping it minimal leads to clearer documentation, easier maintenance and better usability.

smoliji commented:
> > 3. BUT just to be more effective in some cases - you dont want to pass adapters around
> 
> All I tried to say was that this is the 98%. You initialize channel and send. `send` the "real deal" still requires the channel instance which has the method so it does not really bring any advantages.
> 
> I understand the bottom-up construction, but if there is no practical advantage to using one over the other and we all agree that the adapter method is more convenient, why do we need to expose the general `send` method?
> 
> If there is a sensible reason to use `send` directly, explain. If not, I would suggest not including it in the public API. Keeping it minimal leads to clearer documentation, easier maintenance and better usability.

If i should choose between the two, then `send` should be the only one implemented. `sender` is just syntax sugar for bound method to an adapter.

`send` goes well with common FP patterns - library provides you generic functions like [Ramda's path](https://ramdajs.com/docs/#path), but you create your own `getFoo = R.path(['foo'])`.

There is no more sensible reason to use `send` over `sender` and vice versa - they are both equivalent and suits different usecases. If I wanted to send a single message over various adapters, it makes more sense to have a single `send` and map it over array of adapters. But as I stated, ofc you can achieve the same with `sender`.

A greater reason for the `send` is understanding that there is some generic send logic when sending a message. Which might not be obvious enough from just using `adapter.sender` - that its composed from a generic enmail handler and an adapter channel handler.

grissius commented on src/lib/enmail.ts:
```diff
@@ -0,0 +1,16 @@
+
+export interface Adapter<AdapterOptions = any, SendOptions = any> {
+    options: AdapterOptions;
+    mailService: any;
```
There could be a generic param for mailService. It is almost zero-effort and would help greatly when working directly with the instance as shown [here](https://github.com/AckeeCZ/enmail/pull/8/files#diff-333ca5a62d332aeaa095f662a837dee1R51)

stefan-prokop-cz replied:
`MailService` generic param added

grissius commented on src/test/integration/sendgrid.test.ts:
```diff
@@ -0,0 +1,56 @@
+        expect(typeof sender).toEqual('function');
+        expect(sender).toEqual(adapter.sender);
+    });
+    it('An email can be sent via adapter', async () => {
```
This and the following tests just test the signature. It would be cool to test wether options get passed to call somehow (with jests' spy perhaps or module mock), that options overrides work etc. 200 is just nocked.

grissius reviewed (changes requested):
There is deprecated documentation. Please update readme and remove docs to deleted adapters.

stefan-prokop-cz commented:
> There is deprecated documentation. Please update readme and remove docs to deleted adapters.

I would like to update readme in the end - after all services will be implemented. If it's problem I will use sendgrid now.

I added todos to the README file and removed docs of deleted adapters

smoliji reviewed (approved):
Let's see where this is going! :rocket:

grissius reviewed (approved):
Thanks for updates!
